### ISSUE TITLE
Hang when value-to-print is very large

### ISSUE THREAD

**Issue by BenWiederhake** [bpython/bpython#703]
bpython (or maybe curtsies, no idea) hangs when the result is very large.
Example:

```
>>> s = "hello" * 1000
>>> s
[hangs, never returns]
```

This happens on all of the following configurations:
- `bpython version 0.17 on top of Python 2.7.13 /usr/bin/python` on machine A
- `bpython version 0.17 on top of Python 3.5.4rc1 /usr/bin/python3` on machine A
- `bpython version 0.17 on top of Python 2.7.13 /usr/bin/python` on machine B
- `bpython version 0.17 on top of Python 3.5.4rc1 /usr/bin/python3` on machine B

Where:
- Machine A received the bugfix for #607 back then.  I'm not sure if OS updates have overridden that version.
- Machine B never received any manual updates, uses OS version.

It appears to be independent of #607, and I guess it's a regression.

Doing the same in plain python2 or python3 works flawlessly and quickly.

**Comment by sebastinas** [comment]
From #717: Regression introduced in ef91741.

**Comment by sebastinas** [comment]
Pinging @ata2001.

**Comment by ata2001** [comment]
I'm on it.

**Comment by ata2001** [comment]
Okay, `curtsies.formatstring.width_aware_slice()` is slow for some reason on long strings. It becomes much faster when you pass the string in smaller chunks.

For example:

```
print("\x1b[31mhello" * 1000)
```
won't hang because it gets converted to an FmtStr as 1000 chunks instead of one string

```
>>> msg = fmtstr("\x1b[31mhello" * 1000)                                                                                                             [42/1785]
>>> msg
red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+re
d('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red(
'hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('h
ello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hel
lo')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello
')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')
+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+r
ed('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red
('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')+red('hello')...
```

**Comment by ata2001** [comment]
This is definitely a curtsies issue, I'm going to create a PR optimizing `curtsies.formatstring.width_aware_slice()`.

**Comment by BenWiederhake** [comment]
I took the freedom of overtaking you.

**Comment by mock** [comment]
is that fix above the solution to the problem or just a workaround?

**Comment by BenWiederhake** [comment]
It's the actual solution.  At least it takes the part of the code which is responsible for being slow and makes it so fast that it should no longer be the bottleneck.

**Comment by thomasballinger** [comment]
This has been merged and a new release of Curtsies (0.2.12) has @BenWiederhake's patch. This is a big improvement that will help users now (there are at least 4 issues that are dups of this) so big thanks to @BenWiederhake. The behavior is still empirically quadratic 
<img width="726" alt="quadratic" src="https://user-images.githubusercontent.com/458879/36116563-921e469c-0feb-11e8-9980-0665e5a733e2.png">
so still needs some work.

Since this was introduced by `width_aware_slice` in ef91741 we might need to revert that. I don't remember the specific cases this was important for, should figure this out.

**Comment by thomasballinger** [comment]
I've written a linear version of the line splitting behavior that made this still quadratic despite @BenWiederhake fix in curtsies with https://github.com/bpython/curtsies/pull/110 but in the meantime we've stopped using the width-aware solution and released 0.17.1 to fix this behavior.

I'm not totally clear on what width-awareness was solving for us – seems like some edge cases around where to wrap lines that have double-width characters or combining characters in them, but there haven't been any open issues about that. Eventually we can use this for formatting input and correctly position the cursor in the case of double-width and combining characters, but no issues open about that. I'm going to hold off on https://github.com/bpython/bpython/pull/731 until the case for why we need it is clearer and we're confident it doesn't cause any regressions.
